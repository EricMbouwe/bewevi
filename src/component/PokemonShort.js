import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const PokemonShort = ({
  pokemon, handleClick, handleError,
}) => (
  <Link to="/one-pokemon" onClick={() => handleClick(pokemon.id)} className="pokemon-short main">
    <div className="pokemon-short description">
      <div>
        {pokemon.name || 'No Name'}
      </div>
    </div>

    <div className="pokemon-short image-container">
      <img src={`https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${pokemon.id}.png`} className="pokemon-short pokemon-image" alt="pokemon" onError={handleError} />
    </div>
  </Link>
);

PokemonShort.propTypes = {
  pokemon: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  handleClick: PropTypes.func.isRequired,
  handleError: PropTypes.func.isRequired,

};

export default PokemonShort;
