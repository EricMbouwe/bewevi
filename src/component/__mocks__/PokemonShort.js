import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const PokemonShort = ({
  pokemon,
  handleClick,
  handleError,
}) => (
  <>
    {JSON.stringify(pokemon)}
    <img src="#" alt="test" onError={handleError} />
    <Link to="/one-pokemon" onClick={() => handleClick(pokemon.id)}>
      test
    </Link>

  </>
);

PokemonShort.propTypes = {
  pokemon: PropTypes.shape().isRequired,
  handleError: PropTypes.func.isRequired,
  handleClick: PropTypes.func.isRequired,
};

export default PokemonShort;
