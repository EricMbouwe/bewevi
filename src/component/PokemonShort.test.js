import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import ReactTestUtils from 'react-dom/test-utils';
import {
  BrowserRouter, Redirect, Route, Switch,
} from 'react-router-dom';
import PokemonShort from './PokemonShort';

let pokemon;
let handleClick;
let handleError;

beforeEach(() => {
  handleError = jest.fn();
  pokemon = {
    id: '10',
    name: 'raichu',
  };
  handleClick = jest.fn();
});

afterEach(() => {
  handleClick.mockClear();
  handleError.mockClear();
});

describe('<PokemonShort />', () => {
  it('renders a link to one pokemon page', () => {
    render(
      <BrowserRouter>
        <PokemonShort
          pokemon={pokemon}
          handleError={handleError}
          handleClick={handleClick}
        />
        <Redirect to="/" />
        <Switch>
          <Route exact path="/" render={() => <>Main Page</>} />
          <Route exact path={`/one-pokemon/${pokemon.id}`} render={() => <div>One Pokemon Page</div>} />
        </Switch>
      </BrowserRouter>,
    );

    expect(screen.getByText(/Main Page/i)).toBeInTheDocument();
    expect(screen.getByText(/raichu/i)).toBeInTheDocument();
    userEvent.click(screen.getByText(/raichu/i));
    expect(screen.getByText(/One Pokemon Page/i)).toBeInTheDocument();
  });

  it('triggers onError when there are errors of img tags', () => {
    const rendered = ReactTestUtils.renderIntoDocument(
      <BrowserRouter>
        <PokemonShort
          pokemon={pokemon}
          handleError={handleError}
          handleClick={handleClick}
        />
      </BrowserRouter>,
    );
    const imgs = ReactTestUtils.scryRenderedDOMComponentsWithTag(rendered, 'img');
    imgs.map(img => ReactTestUtils.Simulate.error(img));
    expect(handleError.mock.calls.length).toBe(1);
  });

  it('is rendering \'No  Name\' if name of the pokemon doesn\'t exist ', () => {
    pokemon.name = null;
    render(
      <BrowserRouter>
        <PokemonShort
          pokemon={pokemon}
          handleError={handleError}
          handleClick={handleClick}
        />
      </BrowserRouter>,
    );
    expect(screen.getByText(/No Name/i)).toBeInTheDocument();
  });

  it('renders correctly', () => {
    const tree = renderer.create(
      <BrowserRouter>
        <PokemonShort
          pokemon={pokemon}
          handleError={handleError}
        />
      </BrowserRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
