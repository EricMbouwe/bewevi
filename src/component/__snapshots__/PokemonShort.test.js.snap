// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonShort /> renders correctly 1`] = `
<a
  className="pokemon-short main"
  href="/one-pokemon"
  onClick={[Function]}
>
  <div
    className="pokemon-short description"
  >
    <div>
      foo
    </div>
    <div>
      bar
    </div>
  </div>
  <div
    className="pokemon-short image-container"
  >
    <img
      alt="pokemon"
      className="pokemon-short pokemon-image"
      onError={[MockFunction]}
      src="foobar"
    />
  </div>
</a>
`;
